packages/@lwc/engine-core/src/framework/rendering.ts

```diff
@@ -642,6 +655,13 @@ function allocateInSlot(vm: VM, children: VNodes) {
         const vnodes: VNodes = (cmpSlots[slotName] = cmpSlots[slotName] || []);
         ArrayPush.call(vnodes, vnode);
     }
+}
+
+function allocateInSlot(vm: VM, children: VNodes) {
+    const { cmpSlots: oldSlots } = vm;
+    const cmpSlots = (vm.cmpSlots = create(null));
```

Comment:
We create a single SlotSet at the top that'll collect all the slots we find as we traverse.